modules/3d/src/rendering.cpp

```diff
@@ -0,0 +1,364 @@
+}
+
+
+static void drawTriangle(Vec4f verts[3], Vec3f colors[3], Mat& depthBuf, Mat& colorBuf,
+                         bool invDepthMode, ShadingType shadingType)
+{
+    // any of buffers can be empty
+    int width  = std::max(colorBuf.cols, depthBuf.cols);
+    int height = std::max(colorBuf.rows, depthBuf.rows);
+
+    Point minPt(width, height), maxPt(0, 0);
+    for (int i = 0; i < 3; i++)
+    {
+        // round down to cover the whole pixel
+        int x = (int)(verts[i][0]), y = (int)(verts[i][1]);
+        minPt.x = std::min(    x, minPt.x);
+        minPt.y = std::min(    y, minPt.y);
+        maxPt.x = std::max(x + 1, maxPt.x);
+        maxPt.y = std::max(y + 1, maxPt.y);
+    }
+
+    minPt.x = std::max(minPt.x, 0); maxPt.x = std::min(maxPt.x, width);
+    minPt.y = std::max(minPt.y, 0); maxPt.y = std::min(maxPt.y, height);
+
+    Point2f a(verts[0][0], verts[0][1]), b(verts[1][0], verts[1][1]), c(verts[2][0], verts[2][1]);
+    Point2f bc = b - c, ac = a - c;
+    float d = ac.x*bc.y - ac.y*bc.x;
+
+    if (abs(d) < 1e-6)
+    {
+        // degenerated triangle
+        return;
+    }
+
+    float invd = 1.f / d;
+    float invz[3] = { 1.f / verts[0][3], 1.f / verts[1][3], 1.f / verts[2][3] };
+
+    for (int y = minPt.y; y < maxPt.y; y++)
+    {
+        for (int x = minPt.x; x < maxPt.x; x++)
+        {
+            Point2f p(x + 0.5f, y + 0.5f), pc = p - c;
+            // barycentric coordinates
+            Vec3f f;
+            f[0] = ( pc.x * bc.y - pc.y * bc.x) * invd;
+            f[1] = ( pc.y * ac.x - pc.x * ac.y) * invd;
+            f[2] = 1.f - f[0] - f[1];
+            // if inside the triangle
+            if ((f[0] >= 0) && (f[1] >= 0) && (f[2] >= 0))
+            {
+                bool update = false;
+                float zInter;
+                if (!depthBuf.empty())
+                {
+                    zInter = 1.0f / (f[0] * invz[0] + f[1] * invz[1] + f[2] * invz[2]);
+                    float zCurrent = depthBuf.at<float>(y, x);
+                    if (zInter < zCurrent)
+                    {
+                        update = true;
+                        depthBuf.at<float>(y, x) = zInter;
+                    }
+                }
+                else
+                {
+                    update = true;
+                }
+
+                if (!colorBuf.empty() && update)
+                {
+                    Vec3f color {0, 0, 0};
+                    if (shadingType == ShadingType::White)
+                    {
+                        color = { 1.f, 1.f, 1.f };
+                    }
+                    if (shadingType == ShadingType::Flat)
+                    {
+                        color = colors[0];
+                    }
+                    else // ShadingType::Shaded
+                    {
+                        Vec3f interp;
+                        for (int i = 0; i < 3; i++)
+                        {
+                            interp[i] = f[i] * zInter * invz[i];
```

Comment:
`zInter` may be not initialized here.

Reply:
This is an impossible case but it's not evident from the code, fixing it, thanks